src/material/timepicker/timepicker-input.ts

```diff
@@ -0,0 +1,421 @@
+    const currentValue = this.value();
+    const date = this._dateAdapter.parseTime(value, this._dateFormats.parse.timeInput);
+    const hasChanged = !this._dateAdapter.sameTime(date, currentValue);
+
+    if (!date || hasChanged || !!(value && !currentValue)) {
+      // We need to fire the CVA change event for all nulls, otherwise the validators won't run.
+      this._assignUserSelection(date, true);
+    } else {
+      // Call the validator even if the value hasn't changed since
+      // some fields change depending on what the user has entered.
+      this._validatorOnChange?.();
+    }
+  }
+
+  /** Handles the `blur` event. */
+  protected _handleBlur() {
+    const value = this.value();
+
+    // Only reformat on blur so the value doesn't change while the user is interacting.
+    if (value && this._isValid(value)) {
+      this._formatValue(value);
+    }
+
+    this._onTouched?.();
+  }
+
+  /** Handles the `keydown` event. */
+  protected _handleKeydown(event: KeyboardEvent) {
+    // All keyboard events while open are handled through the timepicker.
+    if (this.timepicker().isOpen()) {
+      return;
+    }
+
+    if (event.keyCode === ESCAPE && !hasModifierKey(event) && this.value() !== null) {
+      event.preventDefault();
+      this.value.set(null);
+      this._formatValue(null);
+    } else if ((event.keyCode === DOWN_ARROW || event.keyCode === UP_ARROW) && !this.disabled()) {
+      event.preventDefault();
+      this.timepicker().open();
+    }
+  }
+
+  /** Sets up the code that watches for changes in the value and adjusts the input. */
+  private _respondToValueChanges(): void {
+    effect(() => {
+      const value = this._dateAdapter.deserialize(this.value());
+      const wasValid = this._lastValueValid;
+      this._lastValueValid = this._isValid(value);
+
+      // Reformat the value if it changes while the user isn't interacting.
+      if (!this._hasFocus()) {
+        this._formatValue(value);
+      }
+
+      if (value && this._lastValueValid) {
+        this._lastValidDate = value;
+      }
+
+      // Trigger the validator if the state changed.
+      if (wasValid !== this._lastValueValid) {
+        this._validatorOnChange?.();
+      }
+    });
+  }
+
+  /** Sets up the logic that registers the input with the timepicker. */
+  private _registerTimepicker(): void {
+    effect(() => {
+      const timepicker = this.timepicker();
+      timepicker.registerInput(this);
+      timepicker.closed.subscribe(() => this._onTouched?.());
+      timepicker.selected.subscribe(({value}) => {
+        if (!this._dateAdapter.sameTime(value, this.value())) {
+          this._assignUserSelection(value, true);
+          this._formatValue(value);
+        }
+      });
+    });
+  }
+
+  /** Sets up the logic that adjusts the input if the min/max changes. */
+  private _respondToMinMaxChanges(): void {
+    effect(() => {
+      // Read the min/max so the effect knows when to fire.
```

Comment:
This seems clunky - seems like the effect API is missing a way for users to do this more naturally

Reply:
It's definitely clunky, I've raised it with the folks working on reactivity.